contracts/zero-ex/contracts/src/transformers/FillQuoteTransformer.sol

```diff
@@ -388,31 +392,109 @@ contract FillQuoteTransformer is
-                results.takerTokenSoldAmount =
-                    results.takerTokenSoldAmount.safeAdd(fillResults.takerFeePaid);
+        bytes4 makerAssetProxyId = LibBytesV06.readBytes4(order.makerAssetData, 0);
+        // If it is a Bridge order we fill this directly
+        // rather than filling via 0x Exchange
+        if (makerAssetProxyId == ERC20_BRIDGE_PROXY_ID) {
+            uint256 outputTokenAmount = LibMathV06.getPartialAmountFloor(
+                takerAssetFillAmount,
+                order.takerAssetAmount,
+                order.makerAssetAmount
+            );
+            try
+                this.fillBridgeOrder(order,
+                    takerAssetFillAmount,
+                    outputTokenAmount,
+                    makerToken)
+            returns (FillOrderResults memory fillResults)
+            {
+                results = fillResults;
+            } catch (bytes memory) {
+                // Swallow failures, leaving all results as zero.
+            }
+        } else {
+            // Track changes in the maker token balance.
+            uint256 initialMakerTokenBalance = makerToken.balanceOf(address(this));
+            try
+                exchange.fillOrder
+                    {value: protocolFee}
+                    (order, takerAssetFillAmount, signature)
+                returns (IExchange.FillResults memory fillResults)
+            {
+                // Update maker quantity based on changes in token balances.
+                results.makerTokenBoughtAmount = makerToken.balanceOf(address(this))
+                    .safeSub(initialMakerTokenBalance);
+                // We can trust the other fill result quantities.
+                results.protocolFeePaid = fillResults.protocolFeePaid;
+                results.takerTokenSoldAmount = fillResults.takerAssetFilledAmount;
+                // If the taker fee is payable in the taker asset, include the
+                // taker fee in the total amount sold.
+                if (isTakerFeeInTakerToken) {
+                    results.takerTokenSoldAmount =
+                        results.takerTokenSoldAmount.safeAdd(fillResults.takerFeePaid);
+                }
+            } catch (bytes memory) {
+                // Swallow failures, leaving all results as zero.
             }
-        } catch (bytes memory) {
-            // Swallow failures, leaving all results as zero.
         }
     }
 
+    /// @dev Attempt to fill a ERC20 Bridge order. If the fill reverts,
+    ///      or the amount filled was not sufficient this reverts.
+    /// @param order The bridge order to fill.
+    /// @param inputTokenAmount How much taker asset to fill.
+    /// @param outputTokenAmount How much maker asset to receive.
+    /// @param makerToken The maker token.
+    function fillBridgeOrder(
+        IExchange.Order calldata order,
+        uint256 inputTokenAmount,
+        uint256 outputTokenAmount,
+        IERC20TokenV06 makerToken
```

Comment:
We intentionally keep the expected makerToken here rather than use the derived one in the `makerAssetData`.